### PR TITLE
[WIP] Put EventPipeThread destructor on Thread's destructor

### LINKED ISSUE #168: [Linux][RHEL] Heap contamination detected! HeapFree was called on a heap other than the one that memory was allocated from
https://helix.dot.net/api/2019-06-17/jobs/01c8c73c-d102-4ec4-8680-1e9dd63d454b/workitems/JIT.Stress/console

https://dev.azure.com/dnceng/public/_build/results?buildId=433392&view=logs

```
JIT/Stress/ABI/tailcalls_do/tailcalls_do.sh [FAIL]
      
      Assert failure(PID 24244 [0x00005eb4], Thread: 24253 [0x5ebd]): !"Heap contamination detected! HeapFree was called on a heap other than the one that memory was allocated from.\n" "Possible cause: you used new (executable) to allocate the memory, but didn't use DeleteExecutable() to free it."
          File: /__w/2/s/src/coreclr/src/vm/hosting.cpp Line: 268
          Image: /home/helixbot/work/01c8c73c-d102-4ec4-8680-1e9dd63d454b/Payload/corerun
      
      /home/helixbot/work/01c8c73c-d102-4ec4-8680-1e9dd63d454b/Work/8543e0ef-3fed-493d-b02b-356fdd033e54/Exec/JIT/Stress/ABI/tailcalls_do/tailcalls_do.sh: line 294: 24244 Aborted                 $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
      
      Return code:      1
      Raw output file:      /home/helixbot/work/01c8c73c-d102-4ec4-8680-1e9dd63d454b/Work/8543e0ef-3fed-493d-b02b-356fdd033e54/Exec/JIT/Stress/Reports/JIT.Stress/ABI/tailcalls_do/tailcalls_do.output.txt
      Raw output:
      BEGIN EXECUTION
      /home/helixbot/work/01c8c73c-d102-4ec4-8680-1e9dd63d454b/Payload/corerun tailcalls_do.dll '--tailcalls' '--num-calls' '1000' '--no-ctrlc-summary'
      Stressing tailcalls
      OSVersion: Unix 2.6.32.754
      OSArchitecture: X64
      ProcessArchitecture: X64
      Selecting SysV ABI
      50 callers done (49 successful tailcalls tested)
      100 callers done (98 successful tailcalls tested)
      150 callers done (146 successful tailcalls tested)
      200 callers done (194 successful tailcalls tested)
      250 callers done (244 successful tailcalls tested)
      300 callers done (294 successful tailcalls tested)
      350 callers done (339 successful tailcalls tested)
      400 callers done (388 successful tailcalls tested)
      450 callers done (434 successful tailcalls tested)
      500 callers done (483 successful tailcalls tested)
      550 callers done (532 successful tailcalls tested)
      600 callers done (581 successful tailcalls tested)
      650 callers done (631 successful tailcalls tested)
      700 callers done (681 successful tailcalls tested)
      750 callers done (730 successful tailcalls tested)
      800 callers done (779 successful tailcalls tested)
      850 callers done (828 successful tailcalls tested)
      900 callers done (878 successful tailcalls tested)
      950 callers done (927 successful tailcalls tested)
      1000 callers done (975 successful tailcalls tested)
      975 tailcalls tested
      24 rejected tailcalls. Breakdown:
      [100.00%]: Not enough incoming arg space
      Expected: 100
      Actual: 134
      END EXECUTION - FAILED
```

### PR DESCRIPTION
**Pull request by josalem**
fixes #168 

To work around #168, I'm moving the destructor for `EventPIpeThread` on to the `Thread` object.  This also moves the desctructor from being fired at process exit which can cause issues.

WIP for now to test on CI

### CODE REVIEW

File: src/coreclr/src/vm/threads.cpp [hunk 1]
```diff
@@ -2627,6 +2627,13 @@ Thread::~Thread()
     }
 #endif // FEATURE_EVENT_TRACE
 
+#ifdef FEATURE_PERFTRACING
+    if (EventPipeThread::Get() != nullptr)
```
Comment (jkotas) [inline src/coreclr/src/vm/threads.cpp]: I do not think `~Thread` is guaranteed to be called on its "owning" thread. It is often called on a different thread long after the physical thread that it was attached to died.

**Comment by josalem** [comment]
cc @janvorli 

I'm experimenting with using another `thread_local` struct to do the delete.  I'm not sure that this will actually move the time the delete is called, but may flatten the memory enough to bypass the c++ runtime issue in RHEL6.  Still experimenting...

**Comment by josalem** [comment]
These changes aren't actually correct.  Turns out that the `EventPipeThread` object has to (as it is programmed now) live longer than the physical thread it is attached to, so it is incorrect to delete it at physical thread exit.  Specifically, all `EventPipeThread` objects from _all_ threads will stay alive until a point in time when _all_ EventPipe sessions are disabled.  I'm assuming this has to do with the fact that these objects are used in event writing and we don't want to lose events if a physical thread is very chatty, but short lived.

**Comment by janvorli** [comment]
> Turns out that the `EventPipeThread` object has to (as it is programmed now) live longer than the physical thread it is attached to

How did this work when the holder was a thread local object then? It would be destroyed at each thread exit that way.

**Comment by josalem** [comment]
> How did this work when the holder was a thread local object then? It would be destroyed at each thread exit that way.

The `EventPipeThread` has an array of `EventPipeSessionState`s on it.  Each of those have an `EventPipeThreadHolder` field on them which creates a copy of the `Wrapper` and acts as a hard reference to the underlying memory.  This means that the original `Wrapper` _does_ get deleted on physical thread exit, but its destructor just decrements the internal ref count of the `EventPipeThread` object.  That memory won't be deleted until its ref count hits zero (it calls `delete this` on itself), which won't happen until the `EventPipeBufferManager` calls `DeallocatedBuffers` when _all_ EventPipe sessions have closed.

**Comment by josalem** [comment]
After further reading, the `EventPipeBufferManager` also contains a copy of the `Wrapper`, hence why the destructor for the ostensibly `thread_local` object is getting called on _another_ thread.

**Comment by josalem** [comment]
Closing to go back to the drawing board on how to fix this issue since the original assumptions aren't correct.

**Pull request closed without merging**